### PR TITLE
Update iree_linalg_ext.scatter to have a dimension map

### PR DESCRIPTION
**Pull request by rsuderman**
Indices are not always guaranteed to be descending. Added support for a dimension map that informs the destination of each indice to its destination location.

Includes updates to mhlo.scatter lowering to use this new behavior.

### CODE REVIEW

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 1]
```diff
@@ -85,12 +85,14 @@ def IREELinalgExt_ScatterOp : IREELinalgExt_Op<"scatter",
   let arguments = (ins
       Variadic<AnyRankedTensorOrMemRefType>:$inputs,
       Variadic<AnyRankedTensorOrMemRefType>:$outputs,
+      I64ElementsAttr:$dimension_map,
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:88]: update the description section. I actually don't know if some of context are outdated or not. Updating the outdated part is also welcome.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 2]
```diff
@@ -127,6 +127,26 @@ LogicalResult ScatterOp::verify() {
     return op->emitOpError("expected index depth is static");
   }
 
+  auto dimMapTy = getDimensionMap().getType().cast<ShapedType>();
+  if (dimMapTy.getRank() != 1 || dimMapTy.getNumElements() != indexDepth) {
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:131]: maybe we want `I64ArrayAttr` for dimMap, so we don't have to check if the rank is 1?

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 3]
```diff
@@ -138,6 +140,14 @@ def IREELinalgExt_ScatterOp : IREELinalgExt_Op<"scatter",
       return getUpdateSliceRank() == 0;
     }
 
+    SmallVector<int64_t> dimensionMap() {
+      SmallVector<int64_t> ret;
+      for (const APInt& elem : getDimensionMap()) {
+        ret.push_back(elem.getLimitedValue());
+      }
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:147]: If it is a `I64ArrayAttr`, you can return `extractFromI64ArrayAttr(getDimensionMap())` directly.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 4]
```diff
@@ -127,6 +127,26 @@ LogicalResult ScatterOp::verify() {
...
+    }
+  }
+
+  auto originalType = getOriginalType();
+  for (auto it : llvm::enumerate(dimMap)) {
+    if (it.value() < 0 || it.value() >= originalType.getRank()) {
+      return op->emitOpError("dimension map entry invalid at entry #0")
+             << it.index();
+    }
+  }
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:148]: There is a static method `isInvalid` in this file, we can reuse the method.

(Maybe should rename it more descriptive.)

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 5]
```diff
@@ -336,14 +355,18 @@ LogicalResult ScatterOp::generateScalarImplementation(OpBuilder &b,
...
     Value idx = b.create<memref::LoadOp>(loc, indices(), loadIndices);
     Value cast = b.create<arith::IndexCastOp>(loc, b.getIndexType(), idx);
 
-    if (starts[i])
-      cast = b.create<arith::AddIOp>(loc, cast, starts[i]);
-    starts[i] = cast;
+    auto dim = dimMap[i];
+
+    if (starts[dim])
+      cast = b.create<arith::AddIOp>(loc, cast, starts[dim]);
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:368]: maybe rename `cast` to `ret`.

File: tests/e2e/linalg_ext_ops/scatter.mlir [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                          ins(%update, %indices : tensor<2x2xi32>, tensor<2x2xi32>)
+                          outs(%original : tensor<3x4xi32>) {
+                    ^bb0(%arg0: i32, %arg1: i32):
+                      iree_linalg_ext.yield %arg0 : i32
+  } -> tensor<3x4xi32>
+
+  check.expect_eq_const(%result, dense<[[0, 0, 1, 1], [1, 1, 0, 0], [0, 0, 0, 0]]> : tensor<3x4xi32>) : tensor<3x4xi32>
+
+  return
+}
```
Comment (hanhanW) [inline tests/e2e/linalg_ext_ops/scatter.mlir:129]: github warning: no new line at end of file

**Review by hanhanW — CHANGES_REQUESTED** [review]
overall looks good to me, just few nits.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 1 above)
Comment (rsuderman) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td]: Added the portion about the dimension map. I don't think there is any part out of date but the doc is a little tricky to parse.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 2 above)
Comment (rsuderman) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: Done.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 4 above)
Comment (rsuderman) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: Used.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 3 above)
Comment (rsuderman) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td]: Done.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 5 above)
Comment (rsuderman) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: Done.

File: tests/e2e/linalg_ext_ops/scatter.mlir (hunk 6 above)
Comment (rsuderman) [reply tests/e2e/linalg_ext_ops/scatter.mlir]: Done.

**Review by hanhanW — APPROVED** [review]
LG, thanks!

**Pull request merged**